.NET Micro Framework client

josesimoes opened the pull request:
- rework code to use the same common files as the other platforms
- add support for v4.3 and v4.4
- add example apps for both v4.3 and v4.4
- HTTP fully functional 
- AMQP is a work in progress

azurecla commented:
Hi __@josesimoes__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@josesimoes__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

pierreca commented:
@josesimoes, thanks for the PR, I'll be looking at it with @tameraw. 

Looks like we can't pull as it is, because the folder structure is off, making the change review and merge complicated if not impossible: The code should be under csharp/device, not directly under csharp/. can you pull/merge the latest develop and make sure you fit in the existing folder structure?

also, moving things around and comparing code: 
- Regarding the preprocessor directives: I'd suggest using shorter such as NETMF_43 and NETMF_44. the "F" in MF followed by "FRAMEWORK" and the "V" in the version following "VERSION" feel redundant. If there are no differences between versions 4.3 and 4.4, I'd even suggest we get rid of the version part and just use NETMF.
- there's a conflict in the .gitignore file. very simple to fix when you pull the latest develop
- in SHA.cs, the namespace is inconsistent with the rest of the project, and the hyper-link in the comments is dead
- in NetMF.WebUtility, the namespace is missing
- in Message.cs, looks like most of the properties setter are identical in NETMF and .NET/UWP, I'd suggest to fold the #if/#else inside the getter
- regarding the #if/#else that function differently with string.IsNullOrWhiteSpace: I'd suggest to move the #if/#else inside the extension method, and change the existing code to use the extension method too, that way we avoid having this #if/#else everywhere we need to verify a string. 

That's all I can see right now, it looks very cool! Thanks!
Pierre

josesimoes commented:
Bonjour Pierre! :smile: 

Thank you for your comments. Please see bellow:

1) Folder structure: I got it. For some reason it was off. I've now aligned the folders and file structure.
2) Pre-processor directives. MF_FRAMEWORK_VERSION_VX_Y is the standard directive used in .NETMF projects that's why I've use that one. I've changed most of those to NETMF (which is the other standard directive when only the platform is relevant). There is only one place where the version matters (see method BuildExpiresOn in SharedAccessSignatureBuilder.cs) I had to keep those.
3) Conflict in .gitignore solved.
4) SHA.cs: I kept the original file unchanged on purpose as the code is not mine. I've now changed the namespace to match the rest of the project. The URL is dead indeed. Probably the author has closed the blog. I've removed the link but kept the author name to give it credit for it. 
5) NetMF.WebUtility: added namespace to match the rest of the project.
6) Message.cs properties: moved the #ifdefs to the getter as suggested.
7)  IsNullOrWhiteSpace extension: I've moved it and rework code accordingly. Had to add the new StringExtensions class to WinRT and UWP projects.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/AuthenticationMethodFactory.cs:
```diff
@@ -28,38 +25,23 @@ static internal IAuthenticationMethod GetAuthenticationMethod(IotHubConnectionSt
+#endif
         }
 
-        /// <summary>
```
why are you removing the comments?

josesimoes replied:
Shouldn't... They are back now.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/Common/SHA.cs:
```diff
@@ -0,0 +1,211 @@
+// 
+//  Code by Elze Kool
```
Since Elze Kool wrote it, do we have his authorization or license information on this piece of code? if not we should reach out to him first to make sure it's OK to use his code. looks like he can be found here: https://github.com/elzekool

josesimoes replied:
As far as know there is no license in the code file.
That class has been around since 2009 as far as know. It has been used in countless NETMF projects since.

Will you reach him officially? Or should I do it?

pierreca replied:
If you know him already, personally or online, I'd suggest you do it but if you don't, it doesn't matter if it's you or me. I'll be happy to do it. just let me know :)

josesimoes replied:
Don't know him but I've just sent an email.

pierreca replied:
awesome, thank you!

josesimoes replied:
I've a reply from Elze stating that "Yes it is public domain. Use it as you like. " 
How should I proceed next?

elzekool replied:
If it's to anyone's help, I indeed mailed José the above, so it's okay. Good luck with your project!

pierreca replied:
This is the last open item, everything else looks good. Just to be on the safe side I requested the advice of our opensource licensing/copyright contacts, I'll keep you posted and will be ready to move forward as soon as we get an answer/recommendation.

josesimoes replied:
Awesome! Looking forward to it. :smile:

pierreca replied:
@josesimoes @elzekool Finally got an answer! Apparently this needs to be explicitely added in a header/comments section at the beginning of the file: “Original author Elze Kool has dedicated this code (SHA.cs) to the public domain.” Thanks!

josesimoes replied:
Done!

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithSharedAccessPolicyKey.cs:
```diff
@@ -3,23 +3,20 @@
 {
+    using Microsoft.Azure.Devices.Client.Extensions;
     using System;
-    
```
I would keep these comments everywhere (in the subsequent files too)

josesimoes replied:
Something happen to all those comments... Obviously they belong there and should be kept.

pierreca commented on csharp/device/samples/UWPSample/project.lock.json:
```diff
@@ -11523,7 +11523,7 @@
       ]
     },
     "Microsoft.ApplicationInsights.WindowsApps/1.0.0": {
-      "sha512": "NvBQnFeiFd0O1QdBz06UGApD7zn7ztVi7qO18IsM3EjiXRNgfrEBXB+azNm8XqLY8xGFAqh3HAuSd/wHZMe0XA==",
```
this file and all subsequent files look like autogenerated code that shouldn't be checked in

josesimoes replied:
Those are added by VS as soon as you build the UWP project.
If they are not in the original branch they probably should.
If you don't agree feel free to remove them.

pierreca replied:
If it's autogenerated, there's no need to check them in and review the changes. I'll remove them when we merge the PR.

pierreca commented on csharp/device/iothub_csharp_client.sln:
```diff
@@ -28,6 +28,14 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "NuGet", "NuGet", "{84052DC5
 		NuGet\Microsoft.Azure.Devices.Client.nuspec = NuGet\Microsoft.Azure.Devices.Client.nuspec
 	EndProjectSection
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Azure.Devices.Client.Http.NetMF_44", "Microsoft.Azure.Devices.Client.NetMF\HttpClient_44\Microsoft.Azure.Devices.Client.Http.NetMF_44.csproj", "{1B0C95B7-D1C7-4283-A495-044803FD034A}"
```
Adding the .NET MF projects in the existing solution will break our build and force an additional dependency on the solution (the .NET MF SDK). I would suggest we use a separate solution for the .NET MicroFramework.

josesimoes replied:
Fine with me.
What should I call the solution? 
iothub_csharp_netmf_client?

pierreca replied:
sure, that works :) thanks!

josesimoes replied:
Done!

josesimoes replied:
Still about this removed the original iothub_csharp_client sln file from the PR as it doesn't make sense to be there with this new solution.

Anyway if you are adding the NETMF version to the official SDK you'll have to address the build somewhere...

pierreca replied:
Yes, absolutely, we will address the build question separately.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/Message.cs:
```diff
@@ -128,7 +158,11 @@ public string MessageId
             get
             {
+#if NETMF
+                return (string)(this.GetSystemProperty(MessageSystemPropertyNames.MessageId) ?? string.Empty);
```
isn't this going to throw an InvalidCastException instead of being null? I think the "as" keyword has the desired effect here more than a traditional cast (could be different with the Microframework, I'm not sure)

josesimoes replied:
I've replaced the code as you suggesting it's working and returning a null when the MessageId property doesn't exist.

azurecla commented:
Hi __@josesimoes__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

pierreca commented:
@josesimoes just a quick update on the process: I've pulled the code, I'm rebasing and compressing all the commits into a clean one, then I will push it through our gated build and hopefully it'll pass. if not I'll fix it. I found a couple of other things, questions more than issues, I'm going to add comments in this PR so that we can discuss them here, but there are no actions required on your side. Thanks!

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/Common/Security/SharedAccessSignatureBuilder.cs:
```diff
@@ -15,6 +18,10 @@ namespace Microsoft.Azure.Devices.Client
     using System.Security.Cryptography;
 #endif
+
+#if !NETMF
```
I'm going to remove this

josesimoes replied:
See the next comment. If you remove this and at latter point someone decides to use the crypto API from NETMF they'll have to add the using. Until then it's not necessary so it's OK to remove that.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/Common/Security/SharedAccessSignatureBuilder.cs:
```diff
@@ -104,6 +151,34 @@ static string Sign(string requestString, string key)
+            var hmac = SHA.computeHMAC_SHA256(Convert.FromBase64String(key), Encoding.UTF8.GetBytes(requestString));
+            return Convert.ToBase64String(hmac);
+
+            // computing SHA256 signature using the .NET Micro Framework classes
```
as I understand it it's useful to keep this commented out in case someone wants to switch to another set of (presumably hardware-based) crypto API

josesimoes replied:
You've got it. I think it should be left there for the reason you've pointed out.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClient.cs:
```diff
@@ -139,7 +140,7 @@ public static DeviceClient CreateFromConnectionString(string connectionString, T
                 throw new NotImplementedException();
 #else
-                return new DeviceClient(new AmqpDeviceClient(iotHubConnectionString));
+                //return new DeviceClient(new AmqpDeviceClient(iotHubConnectionString));
```
not sure why this got commented?

pierreca replied:
not sure why this got commented?

josesimoes replied:
This one is easy: the AMQP is still on the works. See the pr message, only the HTTP part is working.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClient.cs:
```diff
@@ -219,19 +250,34 @@ public AsyncTaskOfMessage ReceiveAsync(TimeSpan timeout)
         /// </summary>
         /// <returns>The lock identifier for the previously received message</returns>
+
+#if NETMF
```
For all these methods that are entirely under if/else NETMF blocks I'm thinking about regrouping them all together into a single one, please let me know if you feel it would be a problem.

josesimoes replied:
No problem. I just left each method implementation in NETMF near it's counterpart in the full Framework. It's really a matter on how things should be organized.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClient.cs:
```diff
@@ -261,36 +321,66 @@ public AsyncTask AbandonAsync(Message message)
         /// </summary>
         /// <returns>The task containing the event</returns>
+#if NETMF
+        //public void SendEventBatch(Array<Message> messages)
```
I believe we could uncomment this

josesimoes replied:
Ok

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClientHelper.cs:
```diff
@@ -39,15 +47,28 @@ protected DeviceClientHelper()
+#if NETMF
+        public void Close()
+        {
+
```
just checking that this is supposed/ok to be empty.

josesimoes replied:
Yes. Because there is nothing to close... Left it there in order to keep 100% compatibility with the API.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClientHelper.cs:
```diff
@@ -194,11 +323,24 @@ public Task<Message> ReceiveAsync()
+#if NETMF
+        //public void Task SendEventBatchAsync(IEnumerable<Message> messages)
+        //{
+        //    if (messages == null)
```
throw NotImplementedException instead of commenting out?

josesimoes replied:
Yes

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClientHelper.cs:
```diff
@@ -209,7 +351,14 @@ public Task<Message> ReceiveAsync()
+#if NETMF
+        protected void EnsureOpened(bool explicitOpen)
+        {
+
```
just checking if this should be empty? maybe if it's useless we can remove it altogether?

josesimoes replied:
Again it's there just to keep 100% compatibility with the API.
Giving it some thought I really think it should be left there despite being empty. This way it a developer is working on a project that targets both platforms the code can be exactly the same thus achieving full code reuse. 
Maybe there should be a comment in these methods so that someone revisiting the code doesn't get the impression that something is missing.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/HttpClientHelper.cs:
```diff
@@ -59,7 +90,50 @@ sealed class HttpClientHelper : IHttpClientHelper
+            }
+            else
+            {
+                //this.ExecuteAsync(
```
we could probably change that into some sort of error/exception, or remove it? thoughts?

josesimoes replied:
Throwing an exception would be cleaner... I think...

pierreca replied:
looking at it with fresh eyes: I think the intention was rather to have  try { //code } catch(Exception) { if (throwIfNotFound) throw; }. otherwise if throwifNotFound is false, the request is not even processed.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/HttpClientHelper.cs:
```diff
@@ -123,7 +199,35 @@ sealed class HttpClientHelper : IHttpClientHelper
+        static HttpWebResponse ReadResponseMessageAsync(HttpWebResponse message)
+        {
+            // TODO
+            // unclear what is the purpose of this method and how to implement it
```
that comment makes me a little worried I'll try and figure this out. thoughts welcome!

josesimoes replied:
Nothing to worry about. I latter figured this out but forgot to remove the method call.
In the original code it reads the HTTP response to get the ETag header and the message body (if any).
In NETMF it's more convenient to read the response stream at OnReceive and store the content in the new Message there.
As for the ETag it's read from the respective header at OnReceive too so we are not missing anything here.

pierreca replied:
ok so in that case I believe we can remove the method.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/IotHubConnectionString.cs:
```diff
@@ -35,11 +43,16 @@ public IotHubConnectionString(IotHubConnectionStringBuilder builder)
+#if !WINDOWS_UWP && !NETMF
             this.AmqpEndpoint = new UriBuilder(CommonConstants.AmqpsScheme, builder.HostName, AmqpConstants.DefaultSecurePort).Uri;
+#elif NETMF
+            //this.AmqpEndpoint = "amqps://" + this.HostName + ":" + AmqpConstants.DefaultSecurePort;
```
looks like we could remove this

josesimoes replied:
Please leave it there because it's going to be needed for AMQP

pierreca replied:
I'd rather remove it and it can be added at the same time as AMQP. leaving commented code is confusing because it doesn't say anything about the state of the SDK.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/IotHubConnectionStringBuilder.cs:
```diff
@@ -165,10 +215,37 @@ void Parse(string iotHubConnectionString)
+                Convert.FromBase64String(this.SharedAccessKey);
+            }
+
+            //if (SharedAccessSignatureParser.IsSharedAccessSignature(this.SharedAccessSignature))
```
should we remove this?

josesimoes replied:
Probably yes. I choose not to implement that validation. Not only it's costly to do but it's not necessary. For NETMF implementations the signature is computed in the code so it's doesn't make much sense to me checking something that you've just computed.

pierreca replied:
removed. thanks!

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/Message.cs:
```diff
@@ -144,7 +178,11 @@ public string To
             get
             {
+#if NETMF
+                return (string)(this.GetSystemProperty(MessageSystemPropertyNames.To) ?? string.Empty);
```
I think this is still going to throw instead of returning null if the cast fails, I don't believe it's the desired effect. I'll change these (this one and the following) to "as" casts. thoughts?

josesimoes replied:
Please change it. I did replace all the others as we've discussed. I must have missed this one...

pierreca replied:
changed wherever it could be changed (won't work with non-nullable types such as DateTime)

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithRegistrySymmetricKey.cs:
```diff
@@ -4,11 +4,17 @@
     using System;
+    using Microsoft.Azure.Devices.Client.Extensions;
 
+#if WINDOWS_UWP
```
I'm going to revert these (the DeviceAuthentication* classes can be public even with UWP)

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/DeviceClient.cs:
```diff
@@ -92,12 +98,7 @@ public static DeviceClient Create(string hostname, IAuthenticationMethod authent
         public static DeviceClient CreateFromConnectionString(string connectionString)
         {
-#if WINDOWS_UWP
-            return CreateFromConnectionString(connectionString, TransportType.Http1);
```
why remove this?

josesimoes replied:
A bad merge.
I've commit the correction to these constructors to match the code in develop branch.
Note that the default being AMQP won't work because that is still not implemented.

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/HttpClientHelper.cs:
```diff
@@ -11,32 +21,44 @@ namespace Microsoft.Azure.Devices.Client
     using System.Linq;
     using System.Net;
     using System.Net.Http;
-#if !WINDOWS_UWP
```
is this supposed to be removed?

pierreca commented on csharp/device/Microsoft.Azure.Devices.Client/HttpDeviceClient.cs:
```diff
@@ -262,13 +461,30 @@ protected override Task OnCompleteAsync(string lockToken)
 
             return this.httpClientHelper.DeleteAsync(
-                GetRequestUri(this.deviceId, CommonConstants.DeviceBoundPathTemplate + "/{0}".FormatInvariant(lockToken), null),
+                GetRequestUri(this.deviceId, CommonConstants.DeviceBoundPathTemplate + "/{0}".FormatInvariant(lockToken), new Dictionary<string, string> { { "reject", null } }),
```
is that related to netmf stuff?

josesimoes replied:
Nothing to do with NETMF. Probably a bad merge...
I've check the code in the develop branch and commit the correct line.

pierreca commented:
This has been merged, thanks a lot Jose!
